### ISSUE TITLE
nspawn: uidmap setup hosed

### ISSUE THREAD

**Issue by mbiebl** [systemd/systemd#20989]
Version: 249.4

Running the autopkgtest suite on Debian bookworm/sid, leads to the following failure in boot-and-services:
```
autopkgtest [13:47:40]: test boot-and-services: [-----------------------
none
Created symlink /etc/systemd/system/default.target → /lib/systemd/system/graphical.target.
bash: line 1: 12745 Killed                  /tmp/autopkgtest.HPdyca/build.4W1/src/debian/tests/boot-and-services 2> >(tee -a /tmp/autopkgtest.HPdyca/boot-and-services-stderr >&2) > >(tee -a /tmp/autopkgtest.HPdyca/boot-and-services-stdout)
autopkgtest [13:47:41]: test process requested reboot with marker boot1
none
test_profile (__main__.AppArmorTest)
AppArmor confined unit ... Replacement succeeded for "violator-test".
ok
test_help (__main__.CLITest)
--help works and succeeds ... ok
test_invalid_option (__main__.CLITest)
Calling with invalid option fails ... ok
test_version (__main__.CLITest)
--version works and succeeds ... ok
test_cpushares (__main__.CgroupsTest)
service with CPUShares ... skipped 'test needs to be reworked on unified cgroup hierarchy'
test_simple (__main__.CgroupsTest)
simple service ... skipped 'test needs to be reworked on unified cgroup hierarchy'
test_bash_crash (__main__.CoredumpTest) ... ok
test_log_for_service (__main__.JournalTest) ... ok
test_no_options (__main__.JournalTest) ... ok
test_boot (__main__.NspawnTest) ... ok
test_service (__main__.NspawnTest) ... ERROR
test_failing (__main__.SeccompTest) ... Job for systemd-nspawn@c1.service failed because the control process exited with error code.
See "systemctl status systemd-nspawn@c1.service" and "journalctl -xeu systemd-nspawn@c1.service" for details.
ok
test_0_init (__main__.ServicesTest)
Verify that init is systemd ... ok
test_cron (__main__.ServicesTest) ... ok
test_dbus (__main__.ServicesTest) ... ok
test_gdm3 (__main__.ServicesTest) ... 324 /usr/sbin/gdm3
328 gdm-session-worker [pam/gdm-launch-environment]
363 /usr/libexec/gdm-wayland-session dbus-run-session -- gnome-session --autostart /usr/share/gdm/greeter/autostart
ok
test_logind (__main__.ServicesTest) ... ok
test_network_manager (__main__.ServicesTest) ... ok
test_no_failed (__main__.ServicesTest)
No failed units ... -------- journal for failed service systemd-nspawn@c1.service -----------
-- Journal begins at Tue 2021-10-12 11:45:37 UTC, ends at Tue 2021-10-12 11:48:06 UTC. --
Oct 12 11:48:06 host systemd[1]: Starting Container c1...
Oct 12 11:48:06 host systemd-nspawn[812]: Failed to create directory at /var/lib/machines/c1/usr: Value too large for defined data type
Oct 12 11:48:06 host systemd[1]: systemd-nspawn@c1.service: Main process exited, code=exited, status=1/FAILURE
Oct 12 11:48:06 host systemd[1]: systemd-nspawn@c1.service: Failed with result 'exit-code'.
Oct 12 11:48:06 host systemd[1]: Failed to start Container c1.
FAIL
test_rsyslog (__main__.ServicesTest) ... ok
test_tmp_cleanup (__main__.ServicesTest) ... ok
test_tmp_mount (__main__.ServicesTest) ... ok
test_udev (__main__.ServicesTest) ... ok

======================================================================
ERROR: test_service (__main__.NspawnTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/autopkgtest.HPdyca/build.4W1/src/debian/tests/boot-and-services", line 255, in test_service
    subprocess.check_call(['systemctl', 'start', 'systemd-nspawn@c1'])
  File "/usr/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['systemctl', 'start', 'systemd-nspawn@c1']' returned non-zero exit status 1.

======================================================================
FAIL: test_no_failed (__main__.ServicesTest)
No failed units
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/autopkgtest.HPdyca/build.4W1/src/debian/tests/boot-and-services", line 68, in test_no_failed
    self.assertEqual(failed, [])
AssertionError: Lists differ: ['systemd-nspawn@c1.service loaded failed failed Container c1'] != []

First list contains 1 additional elements.
First extra element 0:
'systemd-nspawn@c1.service loaded failed failed Container c1'

- ['systemd-nspawn@c1.service loaded failed failed Container c1']
+ []

----------------------------------------------------------------------
Ran 23 tests in 6.728s

FAILED (failures=1, errors=1, skipped=2)
autopkgtest [13:48:09]: test boot-and-services: -----------------------]
autopkgtest [13:48:09]: test boot-and-services:  - - - - - - - - - - results - - - - - - - - - -
boot-and-services    FAIL non-zero exit status 1
autopkgtest [13:48:09]: test boot-and-services:  - - - - - - - - - - stderr - - - - - - - - - -
Job for systemd-nspawn@c1.service failed because the control process exited with error code.
See "systemctl status systemd-nspawn@c1.service" and "journalctl -xeu systemd-nspawn@c1.service" for details.
autopkgtest [13:48:10]: @@@@@@@@@@@@@@@@@@@@ summary
boot-and-services    FAIL non-zero exit status 1
qemu-system-x86_64: terminating on signal 15 from pid 796096 (/usr/bin/python3)
```

Running the same test under Debian bullseye does not trigger this failure. I assume this is the reason why this hasn't shown up on our CI infra yet.

I haven't really investigated yet, what causes this failure, but it is at least reproducible.
Does anyone have an idea what update in our dependencies might be the culprit?

**Comment by bluca** [comment]
Are the tests running natively in qemu or through another container layer?

**Comment by mbiebl** [comment]
The kernel certainly differs between a bullseye and bookworm VM: 5.10.x vs 5.14.x.
@poettering are you aware of any changes between 5.10 and 5.14 ?

**Comment by poettering** [comment]
not really, i wonder if we had (have?) some bug in our use of userns mappings in nspawn though.

**Comment by slyon** [comment]
I am seeing the exact same problem in the Ubuntu Jammy autopkgtest, inside qemu, using Linux 5.13.0

**Comment by bluca** [comment]
Any volunteer for bisecting the kernel between 5.10 and 5.13 to figure out what introduced the regression?

**Comment by mbiebl** [comment]
Seems there is no interest in getting this fixed. Should we disable the boot-and-services test in upstream-ci?

**Comment by bluca** [comment]
Would you be able to bisect the kernel to figure out which change introduced the regression? It's hard to understand what to do otherwise

**Comment by mbiebl** [comment]
I don't have time for that, sorry.

**Comment by mrc0mmand** [comment]
I'll try to look into this once I sort out #21589 .

**Comment by mrc0mmand** [comment]
I managed to reproduce it on Arch Linux as well with the same outcome, but a newer kernel:

```
Dec 17 20:43:44 arch.localdomain systemd[1]: Starting Container c1...
Dec 17 20:43:44 arch.localdomain systemd-nspawn[546]: Failed to create directory at /var/lib/machines/c1/usr: Value too large for defined data type
Dec 17 20:43:44 arch.localdomain systemd[1]: systemd-nspawn@c1.service: Main process exited, code=exited, status=1/FAILURE
Dec 17 20:43:44 arch.localdomain systemd[1]: systemd-nspawn@c1.service: Failed with result 'exit-code'.
Dec 17 20:43:44 arch.localdomain systemd[1]: Failed to start Container c1.
```

```
# uname -a
Linux arch.localdomain 5.15.10-arch1-1 #1 SMP PREEMPT Fri, 17 Dec 2021 11:17:37 +0000 x86_64 GNU/Linux
```

**Comment by mrc0mmand** [comment]
Minimal reproducer:

```
# mkdir -p /var/lib/machines/my-container/etc
# cp /etc/os-release /var/lib/machines/my-container/etc/
# systemctl start systemd-nspawn@my-container
Job for systemd-nspawn@my-container.service failed because the control process exited with error code.
See "systemctl status systemd-nspawn@my-container.service" and "journalctl -xeu systemd-nspawn@my-container.service" for details.
# journalctl -u systemd-nspawn@my-container
Dec 17 20:50:44 arch.localdomain systemd[1]: Starting Container my-container...
Dec 17 20:50:44 arch.localdomain systemd-nspawn[644]: Failed to create directory at /var/lib/machines/my-container/usr: Value too large for defined data type
Dec 17 20:50:44 arch.localdomain systemd[1]: systemd-nspawn@my-container.service: Main process exited, code=exited, status=1/FAILURE
Dec 17 20:50:44 arch.localdomain systemd[1]: systemd-nspawn@my-container.service: Failed with result 'exit-code'.
Dec 17 20:50:44 arch.localdomain systemd[1]: Failed to start Container my-container.
```

**Comment by mrc0mmand** [comment]
Also, looking at a similar issue from 2016 (https://github.com/systemd/systemd/pull/4223#issuecomment-253046519) and a following git blame, a possible culprit might be https://github.com/torvalds/linux/commit/ba73d98745be1c10dc3cce68e8d7b95012d07d05#diff-1107c1cca76c74e3df2ad2aee977fef31bb1e9f4821c280a63f3d76e9611c71fR1096. But that's for someone who understands all that fancy stuff there to determine.

**Comment by bluca** [comment]
> Also, looking at a similar issue from 2016 ([#4223 (comment)](https://github.com/systemd/systemd/pull/4223#issuecomment-253046519)) and following a git blame, a possible culprit might be [torvalds/linux@ba73d98#diff-1107c1cca76c74e3df2ad2aee977fef31bb1e9f4821c280a63f3d76e9611c71fR1096](https://github.com/torvalds/linux/commit/ba73d98745be1c10dc3cce68e8d7b95012d07d05#diff-1107c1cca76c74e3df2ad2aee977fef31bb1e9f4821c280a63f3d76e9611c71fR1096). But that's for someone who understands all that fancy stuff there to determine.

Nice find! Fortunately for us, and unfortunately for him, we know the developer who wrote that fancy stuff :-) /cc @brauner - does that error and the minimal reproducer above ring a bell?

**Comment by brauner** [comment]
I'm currently inclined to blame this on a setup failure in `systemd-nspawn`:
```
3295  open_tree(-1, "/var/lib/machines/my-container", OPEN_TREE_CLONE|OPEN_TREE_CLOEXEC) = 8
[...]
3295  syscall_0x1ba(0x8, 0x7f035c824035, 0x9000, 0x7ffdca944ec0, 0x20, 0x1) = 0
[...]
3295  move_mount(8, "", -1, "/var/lib/machines/my-container", MOVE_MOUNT_F_EMPTY_PATH) = 0
[...]
3295  openat(AT_FDCWD, "/var/lib/machines/my-container", O_RDONLY|O_NONBLOCK|O_NOFOLLOW|O_CLOEXEC|O_DIRECTORY) = 8
[...]
3295  mkdirat(8, "root", 0755)          = -1 EOVERFLOW (Value too large for defined data type)
```
In order for an idmapped mount to be created the process `3295` must be privileged. The line
```
3295  syscall_0x1ba(0x8, 0x7f035c824035, 0x9000, 0x7ffdca944ec0, 0x20, 0x1) = 0
```
is the `mount_setattr()` system call which this version of `strace` cannot yet decode.

This creates an idmapped mount with the idmapping that this container uses. So a process running inside the user namespace of the container will be able to create files in there.

However, we can see that in the example above `3295`, the same process that created the idmapped mount - and therefore not running in the user namespace of the container as there's also not `setns(<fd>, CLONE_NEWUSER)` call in the strace output - tries to create a file going through the idmapped mount. But that doesn't work as the `fsuid` and `fsgid` of this process aren't mapped anymore in the idmapped mount. This is why `EOVERFLOW` is returned. So if you want to create files via the idmapped mount it usually needs to be done from within the containers user namespace or it needs to be done by switching to the `fsuid` and `fsgid` of the container in the host.

**Comment by brauner** [comment]
Yeah, this is all because systemd creates all files and directories for empty containers after the creation of the idmapped mount, i.e. it creates:
```
/var/lib/machines/my-container/:
dev  etc  proc  root  run  sys  tmp  usr  var

/var/lib/machines/my-container/dev:

/var/lib/machines/my-container/etc:
localtime  os-release  resolv.conf

/var/lib/machines/my-container/proc:

/var/lib/machines/my-container/root:

/var/lib/machines/my-container/run:

/var/lib/machines/my-container/sys:

/var/lib/machines/my-container/tmp:

/var/lib/machines/my-container/usr:

/var/lib/machines/my-container/var:
log

/var/lib/machines/my-container/var/log:
journal

/var/lib/machines/my-container/var/log/journal:
```
and that can't work. systemd needs to create those files before creating the idmapped mount. And I also see that you're trying to directly exec the systemd binary from the host:
```
execve("/usr/lib/systemd/systemd", ["/usr/lib/systemd/systemd"], 0x564641242b90 /* 9 vars */ <unfinished ...>
```
that won't work eithe. This needs to be done via an idmapped mount as well, i.e. either open the binary and created an idmapped mount just for that file or bind-mount it into the container and create a recursive idmapped mount for the rootfs and then open the binary to exec.

**Comment by mrc0mmand** [comment]
Thanks a lot, @brauner, for the analysis and explanation!

/cc @poettering @keszybz

**Comment by brauner** [comment]
So `outer_child()` does:
```c
remount_idmap(directory, arg_uid_shift, arg_uid_range);
```
and then below:
```c
base_filesystem_create(directory, arg_uid_shift, (gid_t) arg_uid_shift);
```
The call to `base_filesystem_create()` has multiple issues. First, as I mentioned above, it is done after the idmapped mount has been created. Since the caller's `fs{g,u}id` aren't mapped in the idmapped mount this will fail with `EOVERFLOW`. In addition, `base_fileystem_create()` is currently passed `arg_{g,u}id_shift` as arguments but that means you'll be changing the ownership of the files on-disk. But that should not be done when idmapped mounts are used. This is only needed when recursive chown is used.

After that:
```c
        r = mount_all(directory,
                      arg_mount_settings,
                      arg_uid_shift,
                      arg_selinux_apifs_context);
```
is done which creates additional directories and therefore suffers the same problem.

**Comment by brauner** [comment]
The way I see it there are multiple ways to fix it:
1. Similar to bind-mounting to `/run/systemd/nspawn-root` if `/' is used as root directory you could create another `/run/systemd/nspawn-root-unmapped` where you create a bind-mount for the unmapped rootfs that is available after the idmapped mount has been created. The problem with this is obviously that mounts aren't reflected at both mount points and envoking mount propagation to somehow fix this seems overkill.
2. You switch most of those functions to take fd arguments instead of operating on paths which I dislike anyway ;). Then you can have `remount_idmap()` return an fd for the unmapped mount. Then creation requests go through the unampped fd. If you also return an fd for the idmapped mount you can use that to mount via `/proc/self/fd/<nr>`.
3. You split creation of the idmapped mounts into two distinct steps.
    - create the detached idmapped mount
    - attach the mount to the filesystem
   The rest is similar to 2.

**Comment by mbiebl** [comment]
@poettering I guess this needs input from your side

**Comment by poettering** [comment]
(btw, a simply work-around for the test stuff is to precreate all dirs, so that we don#t have to create anything ourselves anymore, the issue should only become visible if you try to boot a minimal tree that lacks some subdirs)

**Comment by poettering** [comment]
Found a nice fix: #22774
